inc/class-session-controller.php

```diff
@@ -121,6 +153,74 @@ public function create_item( $request ) {
 		return $response;
 	}
 
+	/**
+	 * Validate the 2fa related portion of create session reuest.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @param WP_User $user The user to validate the request for.
+	 * @return WP_Error|null
+	 */
+	protected function validate_2fa( $request, WP_User $user ) : ?WP_Error {
+		if ( ! Two_Factor_Core::is_user_using_two_factor( $user->ID ) ) {
+			return null;
+		}
+
+		$user_providers = Two_Factor_Core::get_enabled_providers_for_user( $user );
+		$provider_name_map = [
+			'email' => 'Two_Factor_Email',
+			'totp' => 'Two_Factor_Totp',
+			'backup_codes' => 'Two_Factor_Backup_Codes',
+		];
+
+		$user_providers_public_names = [];
+		foreach ( $user_providers as $provider ) {
+			$user_providers_public_names[] = array_search( $provider, $provider_name_map, true );
+		}
+
+		$error = new WP_Error( '2fa_required', 'Please provide your 2FA code.', [
+			'status' => 401,
+			'2fa_providers' => $user_providers_public_names,
+		] );
+
+		$provider = $provider_name_map[ $request['2fa']['provider'] ];
+
+		// Validate provider / value if it's been passed.
+		if ( ! empty( $request['2fa']['provider'] ) && ! empty( $request['2fa']['value'] ) ) {
+			if ( ! in_array( $provider, $user_providers, true ) ) {
+				$error->add( 'invalid_2fa_provider', 'User does not have this provider enabled.' );
+			}
+			$providers = Two_Factor_Core::get_providers();
+
+			switch ( $provider ) {
```

Comment:
I guess this is directly tied to the providers then? Not great for extensibility, but fine I guess

Reply:
Yes, it's kind of annoying, there's no standard API to validate them, as each provider can output an arbitrary form on the login page, and then it grabs the data from $_POST, so unless all providers add specific method for the REST API or something I don't see a way to support arbitrary providers.